src/wbase/SendChannelShared.cc

```diff
@@ -293,4 +273,153 @@ bool SendChannelShared::_sendBuf(lock_guard<mutex> const& streamLock,
 }
 
 
+bool SendChannelShared::buildAndTransmitError(util::MultiError& multiErr, Task& task, bool cancelled) {
+    auto qId = task.getQueryId();
+    bool scanInteractive = true;
+    waitTransmitLock(*_transmitMgr, scanInteractive, qId);
+    lock_guard<mutex> lock(_tMtx);
+    // Ignore the existing _transmitData object as it is irrelevant now
+    // that there's an error. Create a new one to send the error.
+    TransmitData::Ptr tData = _createTransmit(task);
+    _transmitData = tData;
+    bool largeResult = false;
+    _transmitData->buildDataMsg(task, largeResult, multiErr);
+    LOGS(_log, LOG_LVL_DEBUG, "SendChannelShared::buildAndTransmitError " << _dumpTr());
+    bool lastIn = true;
+    return _prepTransmit(task, cancelled, lastIn);
+}
+
+
+void SendChannelShared::setSchemaCols(Task& task, std::vector<SchemaCol>& schemaCols) {
+    // _schemaCols should be identical for all tasks in this send channel.
+    if (_schemaColsSet.exchange(true) == false) {
+        _schemaCols = schemaCols;
+        // If this is the first time _schemaCols has been set, it is missing
+        // from the existing _transmitData object
+        lock_guard<mutex> lock(_tMtx);
+        if (_transmitData != nullptr) {
+            _transmitData->addSchemaCols(_schemaCols);
+        }
+    }
+}
+
+bool SendChannelShared::buildAndTransmitResult(MYSQL_RES* mResult, int numFields, Task& task, bool largeResult,
+        util::MultiError& multiErr, std::atomic<bool>& cancelled, bool &readRowsOk) {
```

Comment:
What's a reason for passing `cancelled` as a non-const reference? I'm not seeing anything body of this method as well as methods called from here that would modify a value of this atomic. The following should be safer:
```
std::atomic<bool> const& cancelled,
```
My second question in this context is - do you really need to pass into the method the atomic or just its value captured at a time of the call? I have noticed that the loop below is checking:
```
    bool more = true;
    while (more && !cancelled) {
        more = !_transmitData->fillRows(mResult, numFields, tSize);
        ...
```
Could the cancellation condition be changed  (externally) in between iterations of the loop?